exchanges/stream/stream_types.go

```diff
@@ -36,9 +39,44 @@ type ConnectionSetup struct {
 	ResponseCheckTimeout    time.Duration
 	ResponseMaxLimit        time.Duration
 	RateLimit               int64
-	URL                     string
 	Authenticated           bool
 	ConnectionLevelReporter Reporter
+
+	// URL defines the websocket server URL to connect to
+	URL string
+	// Connector is the function that will be called to connect to the
+	// exchange's websocket server. This will be called once when the stream
+	// service is started. Any bespoke connection logic should be handled here.
+	Connector func(ctx context.Context, conn Connection) error
+	// GenerateSubscriptions is a function that will be called to generate a
+	// list of subscriptions to be made to the exchange's websocket server.
+	GenerateSubscriptions func() (subscription.List, error)
+	// Subscriber is a function that will be called to send subscription
+	// messages based on the exchange's websocket server requirements to
+	// subscribe to specific channels.
+	Subscriber func(ctx context.Context, conn Connection, sub subscription.List) error
+	// Unsubscriber is a function that will be called to send unsubscription
+	// messages based on the exchange's websocket server requirements to
+	// unsubscribe from specific channels. NOTE: IF THE FEATURE IS ENABLED.
+	Unsubscriber func(ctx context.Context, conn Connection, unsub subscription.List) error
+	// Handler defines the function that will be called when a message is
+	// received from the exchange's websocket server. This function should
+	// handle the incoming message and pass it to the appropriate data handler.
+	Handler func(ctx context.Context, incoming []byte) error
+}
+
+// ConnectionCandidate contains the connection setup details to be used when
+// attempting a new connection. It also contains the subscriptions that are
+// associated with the specific connection.
+type ConnectionCandidate struct {
```

Comment:
The name Candidate feels wrong to me.
It has Setup, Subscriptions and Connection.
Maybe it's a ConnectionWrapper or something.

Candidate implies something that doesn't feel true; That this is one of several options which may or may not be chosen.

Reply:
Can do, the names I generate are literally the first thing that pops in my head. ask @gloriousCode 😆